### DISCUSSION TITLE
Cant understand why Pyright is giving "reportMissingImports" on local sibling imports

### DISCUSSION THREAD

**Discussion by mlgarrett** [microsoft/pyright#6039] [General]
Hello, I am having a really hard time understanding how to make `Pyright` see my internal imports. My project is structured like this:

```
project
   - folder_a
      a1.py
      a2.py
   - folder_b
      b1.py
      b2.py
   driver_script.py
```

In `b1.py`, for example, I want to use a class from `folder_a`, so I write
```
from folder_a.a1 import Thing
```

Running the `driver_script.py` from the root `project` directory works just fine like this, but in my editor, `Pyright` is telling me that this import cannot be resolved. If I change the import to a relative import, e.g.
```
from ..folder_a.a1 import Thing
```
then `Pyright` is happy, but then running `driver_script.py` throws the exception `Attempted relative import beyond top-level package` and crashes.

I have tried numerous strategies from reading documentation to Google searches, trying to create `pyproject.toml` and/or `pyrightconfig.json`, using `executionEnvironments` and all sorts of other things but I just can't get `Pyright` to be happy with these imports.

Could someone please help me understand what I am missing here? I know this is a common problem but I just can't figure it out.

**Comment by erictraut** [comment]
Converting this from a bug report to a discussion since it's a question.

**Reply by erictraut** [reply]
Are you using the command-line version of pyright? If so, which directory are you running it from?

Or are you using the language server version of pyright? If so, are you opening the project (root) directory within your editor?

I'm not able to repro the problem as it is reported. I've created the directories and files as shown in your example above. Here are the specific file contents I've used:

```python
# a1.py

class Thing: ...
```

```python
# b1.py

from folder_a.a1 import Thing
```

```python
# driver_script.py
from folder_a.a1 import Thing
```

When I run the command-line version of pyright from the root directory, I get the following results:
```
0 errors, 0 warnings, 0 informations
```

Likewise, when I use the pyright language server in VS Code and open the root directory, I see no errors.

![image](https://github.com/microsoft/pyright/assets/7040122/daac193e-2044-41d8-91ab-a3b69d1e5a77)

**Reply by mlgarrett** [reply]
Sorry, I should have specified that I am using the `LSP-pyright` plugin for Sublime Text. And it reports no problems doing the import from the root folder in `driver_script.py`.

But the import line in my `b1.py` shows `reportMissingImports`. If I change that to a relative import,
```
from ..folder_a.a1 import Thing
```
`pyright` is happy. But...

In my project, `b1.py` also contains a class definition, `Thing2`. The `driver_script.py` is importing and using that class:
```
from folder_b.b1 import Thing2
```
and this is where Python throws a `Attempted relative import beyond top-level package` and dies.

So, if I write the import in `b1.py` such that the `driver_script.py` doesn't crash, `pyright` in Sublime Text isn't happy. But if I change to a relative import to make `pyright` happy, running `driver_script.py` crashes.

I hope I am making sense. I really appreciate the advice. Am I doing some kind of unusual/unsanctioned code structure? `b1` imports something from `a1`, and `driver_script` imports something from `b1`.

**Reply by erictraut** [reply]
I'm not able to repro the problem as you've described, at least not in VS Code. I suspect the problem is with your client or the LSP plugin you're using. Based on the symptoms you're describing, it's probably not sending the correct workspace root directory to the pyright language server.

Make sure you're opening the project's folder and not the individual files in the project. (I'm not sure how Sublime works, but VS Code allows you to "Open Folder..." to open a folder as a project.)

If that doesn't fix it, then I recommend contacting the maintainers of the LSP plugin. They should be able to help you diagnose the problem.

In any event, I don't think this is a bug in pyright.

**Reply by mlgarrett** [reply]
Woah, I had no idea about opening the folder. I was always just opening individual files into various Sublime Text windows straight from the command line!

If I `Open Folder` in Sublime Text, nothing **seems** to happen. But then, if I open `b1.py` into that Sublime Text window, `pyright` is totally fine with it! So there really is something special happening when I `Open Folder`?

Thank you so much for your help. I went back to the real project and opened the folder with Sublime Text, resulting in an empty Sublime Text window, and then opened one of the offending files into that window and poof, `pyright` is totally fine. Thanks again, you da best.
